expression/builtin_other.go

```diff
@@ -673,118 +684,354 @@ func (c *setVarFunctionClass) getFunction(ctx sessionctx.Context, args []Express
 	if err = c.verifyArgs(args); err != nil {
 		return nil, err
 	}
-	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETString, types.ETString, types.ETString)
+	argTp := args[1].GetType().EvalType()
+	if argTp == types.ETTimestamp || argTp == types.ETDatetime || argTp == types.ETDuration || argTp == types.ETJson {
+		argTp = types.ETString
+	}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, argTp, types.ETString, argTp)
 	if err != nil {
 		return nil, err
 	}
 	bf.tp.Flen = args[1].GetType().Flen
-	// TODO: we should consider the type of the argument, but not take it as string for all situations.
-	sig = &builtinSetVarSig{bf}
-	return sig, err
+	switch argTp {
+	case types.ETString:
+		sig = &builtinSetStringVarSig{bf}
+	case types.ETReal:
+		sig = &builtinSetRealVarSig{bf}
+	case types.ETDecimal:
+		sig = &builtinSetDecimalVarSig{bf}
+	case types.ETInt:
+		sig = &builtinSetIntVarSig{bf}
+	default:
+		return nil, errors.Errorf("unexpected types.EvalType %v", argTp)
+	}
+	return sig, nil
 }
 
-type builtinSetVarSig struct {
+type builtinSetStringVarSig struct {
 	baseBuiltinFunc
 }
 
-func (b *builtinSetVarSig) Clone() builtinFunc {
-	newSig := &builtinSetVarSig{}
+func (b *builtinSetStringVarSig) Clone() builtinFunc {
+	newSig := &builtinSetStringVarSig{}
 	newSig.cloneFrom(&b.baseBuiltinFunc)
 	return newSig
 }
 
-func (b *builtinSetVarSig) evalString(row chunk.Row) (res string, isNull bool, err error) {
+func (b *builtinSetStringVarSig) evalString(row chunk.Row) (res string, isNull bool, err error) {
 	var varName string
 	sessionVars := b.ctx.GetSessionVars()
 	varName, isNull, err = b.args[0].EvalString(b.ctx, row)
 	if isNull || err != nil {
 		return "", isNull, err
 	}
-
 	datum, err := b.args[1].Eval(row)
 	isNull = datum.IsNull()
 	if isNull || err != nil {
 		return "", isNull, err
 	}
-
 	res, err = datum.ToString()
 	if err != nil {
 		return "", isNull, err
 	}
-
 	varName = strings.ToLower(varName)
 	sessionVars.UsersLock.Lock()
 	sessionVars.SetUserVar(varName, stringutil.Copy(res), datum.Collation())
 	sessionVars.UsersLock.Unlock()
 	return res, false, nil
 }
 
+type builtinSetRealVarSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinSetRealVarSig) Clone() builtinFunc {
+	newSig := &builtinSetRealVarSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinSetRealVarSig) evalReal(row chunk.Row) (res float64, isNull bool, err error) {
+	var varName string
+	sessionVars := b.ctx.GetSessionVars()
+	varName, isNull, err = b.args[0].EvalString(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, err
+	}
+	datum, err := b.args[1].Eval(row)
+	isNull = datum.IsNull()
+	if isNull || err != nil {
+		return 0, isNull, err
+	}
+	res = datum.GetFloat64()
+	varName = strings.ToLower(varName)
+	sessionVars.UsersLock.Lock()
+	sessionVars.Users[varName] = datum
+	sessionVars.UsersLock.Unlock()
```

Comment:
How about creating a new method `SetUsersVar(var, type)` on `sessionVars` for this logic, which makes it safer and more easy-to-use?

Reply:
If the caller has no consciousness that this access should be protected by lock, then even if we wrap this logic as a function, they would not be aware of its existence?